File: scala/secret-handshake/src/main/scala/SecretHandshake.scala
object SecretHandshake {
  private[this] val COMMANDS: List[(Int, String)] = List(
    1 -> "wink",
    2 -> "double blink",
    4 -> "close your eyes",
    8 -> "jump"
  )

  private[this] val REVERSE_COMMAND = 16

  def commands(n: Int): List[String] = {
    val result = COMMANDS.flatMap(c => if ((n & c._1) != 0) Option.apply(c._2) else None)
    if ((n & REVERSE_COMMAND) != 0)
      result.reverse
    else
      result
  }
}

File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits
import Data.Maybe

handshake :: Int -> [String]
handshake n = if testBit n reverseSignal then reverse actionList else actionList

    where
        actions = [(0, "wink"), (1, "double blink"), (2, "close your eyes"), (3, "jump")]
        reverseSignal = 4
        getAction x = if testBit n $ fst x then Just $ snd x else Nothing
        actionList = [action | action <- mapMaybe getAction actions]
